Development/OpenPassDevelopmentApp/DeviceAuthorizationViewModel.swift

```diff
@@ -0,0 +1,126 @@
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+//
+
+#if os(tvOS)
+
+import CoreImage.CIFilterBuiltins
+import Foundation
+import OpenPass
+import SwiftUI
+
+extension DeviceAuthorizationViewModel {
+
+    /// A interface defining the flow of state communicated by the `DeviceAuthorizationFlow`
+    public enum State: Sendable {
+
+        /// The client has been initialized but a ``DeviceCode`` has not been requested or received
+        case initial
+
+        case deviceCodeAvailable(DeviceCode)
+
+        /// The previous ``DeviceCode`` has now expired, and the consumer is required to re-start the flow via
+        /// ``DeviceAuthorizationFlow.fetchDeviceCode()``.
+        case deviceCodeExpired
+
+        /// An unexpected error has occurred.
+        case error(Error)
+
+        /// The flow is complete and the associated ``OpenPassManager`` has obtained the set of ``OpenPassTokens``.
+        case complete(OpenPassTokens)
+    }
+}
+
+@MainActor
+final class DeviceAuthorizationViewModel: ObservableObject {
+
+    @Published private(set) var deviceCode: DeviceCode?
+
+    @Published private(set) var verificationUriCompleteImage: UIImage?
+
+    @Published private(set) var error: Error?
+
+    @Published private(set) var state: DeviceAuthorizationViewModel.State = .initial {
+        didSet {
+            switch state {
+            case .initial:
+                break
+            case .deviceCodeAvailable(let deviceCode):
+                self.deviceCode = deviceCode
+                if let verificationUriComplete = deviceCode.verificationUriComplete {
+                    verificationUriCompleteImage = qrCodeImage(url: verificationUriComplete)
+                }
+            case .deviceCodeExpired:
+                self.deviceCode = nil
+            case .error(let error):
+                self.error = error
+            case .complete:
+                break
+            }
+        }
+    }
+
+    private var signInTask: Task<Void, Never>?
+
+    public func startSignInDAFFlow() {
+        signInTask = Task {
+            let flow = OpenPassManager.shared.deviceAuthorizationFlow
+
+            do {
+                // Request a Device Code
+                let deviceCode = try await flow.fetchDeviceCode()
+                self.state = .deviceCodeAvailable(deviceCode)
+
+                // Poll for authorization
+                let tokens = try await flow.fetchAccessTokenPolling(deviceCode: deviceCode)
+                self.state = .complete(tokens)
+            } catch {
+                self.state = .error(error)
+            }
+        }
+    }
+
+    public func cancelSignIn() {
+        signInTask?.cancel()
+        signInTask = nil
```

Comment:
After a task has been cancelled, should we reset `self.state`? I'm just thinking about the scenario where it's polling for authorization, so the state is `deviceCodeAvailable`. Then we cancel, but that old DeviceCode is still being Published, right?